[ZEPPELIN-3623] Create interpreter process in the cluster mode

liuxunorg opened the pull request:
### What is this PR for?
In distributed cluster deployment mode, look for servers with idle resources in the cluster and create an interpreter process. In single-server deployment mode, the interpreter process is created directly on the machine.

Just set the cluster server list in `zeppelin-site.xml`
```
<property>
  <name>zeppelin.cluster.addr</name>
  <value>10.120.196.234:6000,10.120.196.235:6000,10.120.196.236:6000</value>
</property>
```
The interpreter process can be created in the cluster server through any zeppelin server.

[Cluster Design Document](https://docs.google.com/document/d/1a8QLSyR3M5AhlG1GIYuDTj6bwazeuVDKCRRBm-Qa3Bw/edit#heading=h.s41ckl271z8s)

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3623

### How should this be tested?
* [CI Pass](https://travis-ci.org/liuxunorg/zeppelin/builds/537437337)

### Screenshots (if appropriate)

#### The note is executed on the zeppelin-server-234 and the interpreter process is created on the zeppelin-server-236 host.

![ClusterCreateIntpProcess](https://user-images.githubusercontent.com/3677382/58382741-7dc61680-8000-11e9-8379-78b8e6743b63.gif)


### Questions:
* Does the licenses files need update?
* Is there breaking changes for older versions?
* Does this needs documentation?

felixcheung commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/cluster/ClusterManagerServer.java:
```diff
@@ -312,8 +315,24 @@ public boolean openRemoteInterpreterProcess(
+    } catch (InterpreterNotFoundException e) {
+      e.printStackTrace();
+    } catch (InterpreterException e) {
+      e.printStackTrace();
```
use LOGGER.error, don't use e.printStrackTrace()

felixcheung commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java:
```diff
@@ -144,6 +162,73 @@ public static synchronized ZeppelinConfiguration create() {
+    try {
+      realPath = java.net.URLDecoder.decode(realPath, "utf-8");
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```
why wrap this exception?

felixcheung commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/ManagedInterpreterGroup.java:
```diff
@@ -49,12 +60,91 @@
+            properties.put(ClusterMeta.INTP_TSERVER_PORT, intpTSrvPort);
+          }
+        } catch (TException e) {
+          LOGGER.error(e.getMessage());
```
LOGGER.error(message, e)

liuxunorg commented on zeppelin-interpreter-api/pom.xml:
```diff
@@ -68,8 +68,6 @@
               <exclude>org.apache.commons:commons-exec</exclude>
               <!-- Leave log4j unshaded so downstream users can configure logging. -->
               <exclude>log4j:log4j</exclude>
-              <exclude>com.esotericsoftware:kryo</exclude>
```
@zjffdu , I see that you excluded `kryo` in https://github.com/apache/zeppelin/commit/ae7318f4c443a1e0fb1e6bc4758ad99092b880c4.
But the cluster needs `kryo`, So I shaded and rename `kryo` in interpreter-api pom.
Please help me review here, See if it will affect the operation of spark, Thanks you!

zjffdu replied:
Let's see if this affects the test ZeppelinSparkClusterTest.java

liuxunorg replied:
ok, [CI Pass](https://travis-ci.org/liuxunorg/zeppelin/builds/537437337), ZeppelinSparkClusterTest.java no problem.

liuxunorg commented:
Will merge if no more comments

zjffdu commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java:
```diff
@@ -119,6 +121,25 @@ public static synchronized ZeppelinConfiguration create() {
     }
 
+    // Support for using ZeppelinConfiguration in RemoteInterpreterServer
+    if (url == null) {
```
Why add this piece of code ? What happens without it ?

liuxunorg replied:
Because `RemoteInterpreterServer.java` needs to get the configuration information of zconf.
Obtain the configuration parameters of the cluster and connect to the cluster.
```
<property>
  <name>zeppelin.cluster.addr</name>
  <value>10.120.196.234:6000,10.120.196.235:6000,10.120.196.236:6000</value>
</property>
```

zjffdu replied:
Sorry, I didn't get it. Why RemoteInterpreterServer could not get `zeppelin-site.xml` ？ It should be launched in any host of cluster, and ZEPPELIN_HOME, ZEPPELIN_CONF_DIR should be passed to interpreter process properly.

liuxunorg replied:
Ok, I modified to get the `zeppelin-site.xml` path from the environment variable `ZEPPELIN_HOME` or `ZEPPELIN_CONF_DIR `.

zjffdu commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/Interpreter.java:
```diff
@@ -53,6 +54,24 @@
+  //    this flag in zepplin-server B is false, and zepplin-server B creates intp process locally.
+  protected boolean createInptProcessInCluster = false;
+
+  public void setCreateInptProcessInCluster(boolean createInptProcessInCluster) {
```
I don't think it is proper to put interpreter launch logic in Interpreter itself, it should be the interpreter launch component to handle it.

liuxunorg replied:
Ok, accept your opinion, Let me solve this problem again.

zjffdu commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterServer.java:
```diff
@@ -178,6 +187,8 @@ public RemoteInterpreterServer(String intpEventServerHost,
         new TThreadPoolServer.Args(serverTransport).processor(processor));
     remoteWorksResponsePool = Collections.synchronizedMap(new HashMap<String, Object>());
+
+    clusterManagerClient.start(interpreterGroupId);
```
Even in single node deployment, this code is still called ?

liuxunorg replied:
All interpreter processes report their status to the cluster.
After the zeppelin cluster has metadata for all zeppelin services and interpreter processes,
It will be of great help to the management of zeppelin in the future.
E.g,
1. Cluster metadata can know which interpreter processes are started on which servers.
2. The operating state of each interpreter process,
3. CPU and memory resource usage for each interpreter process.
4. Even the user's session information can be managed uniformly.

Therefore, I consider that the single machine is also managed by cluster metadata.

zjffdu commented:
@liuxunorg Thanks for the PR. I left a few comments. Overall I don't think it is proper to do it in injection approach. A lots of cluster deployment specific logic is injected into Interpreter, RemoteInterpreterServer and etc. I think cluster deployment is another kind of launch approach, we should move them out of the standard interpreter classes to make them decouple of these basic classes.

liuxunorg commented:
@zjffdu , I re-implemented it again, The logic is much clearer.
1. Added `plugins/ClusterLauncher`
2. Create an interpreter process on other zeppelin servers by means of message broadcasts.
3. Remove the old way to create interpreter processes on other zeppelin servers via the thrift interface.
4. Added `ClusterManagerTest` , `ClusterMultiNodeTest` and `ClusterMockTest` test case.

Please help review, Thank you!

zjffdu commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/cluster/event/ClusterEvent.java:
```diff
@@ -0,0 +1,24 @@
+ * Cluster Event
+ */
+public enum ClusterEvent {
+  CREATE_INTP_PROCESS
```
Why other events not in this enum ? Like NEW_NOTE、DEL_NOTE、REMOVE_NOTE_TO_TRASH and etc.

liuxunorg replied:
Other types, Will be added in the next PR.
https://issues.apache.org/jira/browse/ZEPPELIN-3778

liuxunorg commented:
Any more comments?  :-)

zjffdu commented:
LGTM

zjffdu commented:
@liuxunorg Sorry, the CI is failed, could you check that ?

liuxunorg commented:
[CI Pass](https://travis-ci.org/liuxunorg/zeppelin/builds/543305876)
Will merge if no more comments

liuxunorg commented:
[CI failure is unrelated.](https://travis-ci.org/liuxunorg/zeppelin/builds/545043371)
I have developed other PR, relying on this PR.
Will merge this if there are no more discussions.
